packages/theme/src/cli/utilities/theme-uploader.ts

```diff
@@ -32,82 +33,183 @@ export async function uploadTheme(
-    task: async () => {
-      await deleteThemeAsset(theme.id, file.key, session)
-    },
-  }))
+  const progress = {current: 0, total: orderedFiles.length}
+  const promise = Promise.all(
+    orderedFiles.map((file) =>
+      deleteThemeAsset(theme.id, file.key, session).then(() => {
+        progress.current++
+      }),
+    ),
+  ).then(() => {
+    progress.current = progress.total
+  })
+
+  return {progress, promise}
 }
 
 async function getRemoteFilesToBeDeleted(
   remoteChecksums: Checksum[],
   themeFileSystem: ThemeFileSystem,
   options: UploadOptions,
 ): Promise<Checksum[]> {
-  const localKeys = new Set(themeFileSystem.files.keys())
   const filteredChecksums = await applyIgnoreFilters(remoteChecksums, themeFileSystem, options)
-  const filesToBeDeleted = filteredChecksums.filter((checksum) => !localKeys.has(checksum.key))
+  const filesToBeDeleted = filteredChecksums.filter((checksum) => !themeFileSystem.files.has(checksum.key))
   outputDebug(`Files to be deleted:\n${filesToBeDeleted.map((file) => `-${file.key}`).join('\n')}`)
   return filesToBeDeleted
 }
 
 // Contextual Json Files -> Json Files -> Liquid Files -> Config Files -> Static Asset Files
 function orderFilesToBeDeleted(files: Checksum[]): Checksum[] {
-  const {contextualizedJsonFiles, jsonFiles, liquidFiles, configFiles, staticAssetFiles} = partitionThemeFiles(files)
-  return [...contextualizedJsonFiles, ...jsonFiles, ...liquidFiles, ...configFiles, ...staticAssetFiles]
+  const fileSets = partitionThemeFiles(files)
+  return [
+    ...fileSets.contextualizedJsonFiles,
+    ...fileSets.templateJsonFiles,
+    ...fileSets.otherJsonFiles,
+    ...fileSets.sectionLiquidFiles,
+    ...fileSets.otherLiquidFiles,
+    ...fileSets.configFiles,
+    ...fileSets.staticAssetFiles,
+  ]
+}
+
+interface SyncJob {
+  progress: {current: number; total: number}
+  promise: Promise<void>
 }
 
-async function buildUploadTasks(
+async function buildUploadJob(
   remoteChecksums: Checksum[],
   themeFileSystem: ThemeFileSystem,
   options: UploadOptions,
   theme: Theme,
   session: AdminSession,
   uploadResults: Map<string, Result>,
-): Promise<Task[]> {
+): Promise<SyncJob> {
   const filesToUpload = await selectUploadableFiles(themeFileSystem, remoteChecksums, options)
-  await readThemeFilesFromDisk(filesToUpload, themeFileSystem)
-  return createUploadTasks(filesToUpload, themeFileSystem, session, theme, uploadResults)
+  const orderedFiles = orderFilesToBeUploaded(filesToUpload)
+
+  const progress = {current: 0, total: filesToUpload.length}
+  const promise = Promise.all(
+    orderedFiles.flatMap((fileType) => {
+      if (fileType.length === 0) return []
+
+      return createBatches(fileType).map((batch) => {
+        return uploadBatch(batch, themeFileSystem, session, theme.id, uploadResults).then(() => {
+          progress.current += batch.length
+        })
+      })
+    }),
+  ).then(() => {
+    progress.current = progress.total
+  })
```

Comment:
When we flat map the files, I wonder if we're exposing the logic of the order of files, as requests for Liquid sections and JSON templates are triggered simultaneously.

The files with dependencies we have are:

1. Liquid sections need to be uploaded first so they may be referenced by templates.
2. JSON sections need to be uploaded afterward so they can reference Liquid sections.
3. JSON templates should follow so they can reference sections.
4. Contextualized templates can be uploaded after templates.
5. Config files must be the last ones, but we need to upload config/settings_schema.json first, followed by config/settings_data.json.

The files with no dependencies we have are:

- The other Liquid files (for example, snippets).
- The other JSON files (for example, locales).
- The static assets.

We do have the `handleFailedUploads` function that retries 2 times if something goes wrong. However, given the variability in theme sizes (some themes have very large Liquid files, others have large JSON files), I'm concerned that if we don't guarantee the dependency order, the retry mechanism might not be enough to ensure successful uploads.